clientv3: check nil connection in lease

gyuho opened the pull request:
Fix https://github.com/coreos/etcd/issues/5527.

gyuho commented:
/cc @heyitsanthony @xiang90 Please review.

Thanks.

xiang90 commented:
For #5527, where do we stop the conn? I think it panicked even before we stopped it?

gyuho commented:
@xiang90 Right. I will investigate more in server side.

xiang90 commented:
@gyuho No. It was the client that panicked.

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -394,6 +411,13 @@ func (l *lessor) getKeepAliveStream() pb.Lease_LeaseKeepAliveClient {
 func (l *lessor) switchRemoteAndStream(prevErr error) error {
 	for {
+		l.rc.mu.Lock()
+		closed := l.rc.client.cancel == nil
```
this probably needs to use `rc.acquire()`, otherwise a `client.Close()` can happen immediately after the `Unlock()` and the `closed` value will be out of date

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -163,6 +170,13 @@ func (l *lessor) Revoke(ctx context.Context, id LeaseID) (*LeaseRevokeResponse,
 }
 
 func (l *lessor) KeepAlive(ctx context.Context, id LeaseID) (<-chan *LeaseKeepAliveResponse, error) {
+	l.rc.mu.Lock()
```
I don't think this belongs here since the method delegates the client connection handling to the goroutine

gyuho replied:
right. will rework on this. thanks.

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -119,6 +119,13 @@ func (l *lessor) Grant(ctx context.Context, ttl int64) (*LeaseGrantResponse, err
 
 	for {
+		l.rc.mu.Lock()
+		closed := l.rc.client.cancel == nil
```
this check isn't safe either-- if the client rlock isn't held for the span of checking if the client is closed and using the grpc LeaseClient then a Close can still happen in between and break it

gyuho commented:
@xiang90 @heyitsanthony Can you take a look again? 
- Changed it to use `acquire`
- Changed it to check `closed` inside `keepAliveCtx` goroutine

Not sure about the `Grant` part yet. @heyitsanthony do you have any suggestion to check `closed` inside `Grant` method?

Btw, this is not fixing https://github.com/coreos/etcd/issues/5527. This is for the similar issues that we had with KV and watch API. I will investigate more the LeasePromote integration test failure, along with this.

Thanks.

gyuho commented on clientv3/lease.go:
```diff
@@ -119,6 +119,13 @@ func (l *lessor) Grant(ctx context.Context, ttl int64) (*LeaseGrantResponse, err
 
 	for {
+		l.rc.client.mu.RLock()
+		closed := l.rc.client.cancel == nil
```
Anthony pointed out

> this check isn't safe either-- if the client rlock isn't held for the span of checking if the client is closed and using the grpc LeaseClient then a Close can still happen in between and break it

heyitsanthony commented:
@gyuho I think this isn't necessary as of 7dfe7db2431bdfe1247cbc5fbc2efabaed67aae6 now that ActiveConnection isn't allowed to go to nil anymore?

gyuho commented:
@heyitsanthony Right. Can you add more tests? If it's not needed, we can skip the tests too.

Thanks. Closing.
